File: user_pathing_creator.py
import sys
import json
from PyQt5.QtWidgets import QApplication, QMainWindow, QFileDialog, QGraphicsScene, QGraphicsView, QGraphicsEllipseItem, QGraphicsItem, QVBoxLayout, QWidget, QPushButton, QComboBox
from PyQt5.QtCore import Qt, QPointF
from PyQt5.QtGui import QBrush

# Constants for UI
ICON_SIZE = 10

class PathEditor(QGraphicsView):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.scene = QGraphicsScene()
        self.setScene(self.scene)
        self.path_points = []  # Store points for editing paths

    def load_path(self, data):
        """Load path data from JSON and display it"""
        self.scene.clear()
        self.path_points = []
        
        for point in data.get("path", []):
            x, y = point["x"], point["y"]
            capture_type = point.get("capture_type", "Default")
            icon = CapturePoint(x, y, capture_type)
            self.scene.addItem(icon)
            self.path_points.append(icon)
    
    def save_path(self):
        """Save current path as JSON"""
        path_data = [{"x": icon.x(), "y": icon.y(), "capture_type": icon.capture_type} for icon in self.path_points]
        return {"path": path_data}

class CapturePoint(QGraphicsEllipseItem):
    def __init__(self, x, y, capture_type):
        super().__init__(0, 0, ICON_SIZE, ICON_SIZE)
        self.setPos(x, y)
        self.setBrush(QBrush(Qt.red))
        self.setFlag(QGraphicsItem.ItemIsMovable)
        self.setFlag(QGraphicsItem.ItemIsSelectable)
        self.capture_type = capture_type

    def mousePressEvent(self, event):
        if event.button() == Qt.LeftButton:
            self.show_property_editor()
        super().mousePressEvent(event)

    def show_property_editor(self):
        """Show dropdown menu to edit capture type"""
        editor = CapturePointEditor(self)
        editor.show()

class CapturePointEditor(QWidget):
    def __init__(self, capture_point):
        super().__init__()
        self.capture_point = capture_point
        self.setWindowTitle("Edit Capture Point")
        self.setGeometry(100, 100, 200, 100)
        
        layout = QVBoxLayout()
        
        self.dropdown = QComboBox()
        self.dropdown.addItems(["Default", "Image Scan", "Video Capture", "Depth Scan"])
        self.dropdown.setCurrentText(self.capture_point.capture_type)
        layout.addWidget(self.dropdown)
        
        save_button = QPushButton("Save")
        save_button.clicked.connect(self.save_capture_type)
        layout.addWidget(save_button)
        
        self.setLayout(layout)
    
    def save_capture_type(self):
        self.capture_point.capture_type = self.dropdown.currentText()
        self.close()

class MainWindow(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("User Path Editor")
        self.setGeometry(100, 100, 800, 600)
        
        self.editor = PathEditor()
        self.setCentralWidget(self.editor)
        
        # Menu Bar
        menu = self.menuBar()
        file_menu = menu.addMenu("File")
        
        load_action = file_menu.addAction("Load Path")
        load_action.triggered.connect(self.load_path)
        
        save_action = file_menu.addAction("Save Path")
        save_action.triggered.connect(self.save_path)
    
    def load_path(self):
        file_name, _ = QFileDialog.getOpenFileName(self, "Open Path File", "", "JSON Files (*.json)")
        if file_name:
            with open(file_name, "r") as file:
                data = json.load(file)
                self.editor.load_path(data)
    
    def save_path(self):
        file_name, _ = QFileDialog.getSaveFileName(self, "Save Path File", "", "JSON Files (*.json)")
        if file_name:
            with open(file_name, "w") as file:
                json.dump(self.editor.save_path(), file, indent=4)

if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = MainWindow()
    window.show()
    sys.exit(app.exec_())

File: user_pathing_creator.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Text.Json;
using System.Windows.Forms;

namespace PathEditor
{
    public partial class MainForm : Form
    {
        private List<CapturePoint> pathPoints = new List<CapturePoint>();
        private CapturePoint selectedPoint = null;

        public MainForm()
        {
            InitializeComponent();
            this.DoubleBuffered = true;
        }

        private void MainForm_Paint(object sender, PaintEventArgs e)
        {
            Graphics g = e.Graphics;
            foreach (var point in pathPoints)
            {
                point.Draw(g);
            }
        }

        private void MainForm_MouseDown(object sender, MouseEventArgs e)
        {
            foreach (var point in pathPoints)
            {
                if (point.Contains(e.Location))
                {
                    selectedPoint = point;
                    if (e.Button == MouseButtons.Left)
                    {
                        ShowPropertyEditor(selectedPoint);
                    }
                    return;
                }
            }
        }

        private void MainForm_MouseMove(object sender, MouseEventArgs e)
        {
            if (selectedPoint != null && e.Button == MouseButtons.Right)
            {
                selectedPoint.Position = e.Location;
                Invalidate();
            }
        }

        private void MainForm_MouseUp(object sender, MouseEventArgs e)
        {
            selectedPoint = null;
        }

        private void ShowPropertyEditor(CapturePoint point)
        {
            using (PropertyEditor editor = new PropertyEditor(point))
            {
                editor.ShowDialog();
                Invalidate();
            }
        }

        private void LoadPath(string filePath)
        {
            string json = File.ReadAllText(filePath);
            var data = JsonSerializer.Deserialize<PathData>(json);
            pathPoints.Clear();

            foreach (var p in data.Path)
            {
                pathPoints.Add(new CapturePoint(new PointF(p.X, p.Y), p.CaptureType));
            }

            Invalidate();
        }

        private void SavePath(string filePath)
        {
            var data = new PathData();
            foreach (var point in pathPoints)
            {
                data.Path.Add(new PathPoint(point.Position.X, point.Position.Y, point.CaptureType));
            }

            string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(filePath, json);
        }

        private void OpenFile_Click(object sender, EventArgs e)
        {
            OpenFileDialog openFileDialog = new OpenFileDialog { Filter = "JSON Files (*.json)|*.json" };
            if (openFileDialog.ShowDialog() == DialogResult.OK)
            {
                LoadPath(openFileDialog.FileName);
            }
        }

        private void SaveFile_Click(object sender, EventArgs e)
        {
            SaveFileDialog saveFileDialog = new SaveFileDialog { Filter = "JSON Files (*.json)|*.json" };
            if (saveFileDialog.ShowDialog() == DialogResult.OK)
            {
                SavePath(saveFileDialog.FileName);
            }
        }
    }

    public class CapturePoint
    {
        public PointF Position { get; set; }
        public string CaptureType { get; set; }
        private static readonly int SIZE = 10;

        public CapturePoint(PointF position, string captureType)
        {
            Position = position;
            CaptureType = captureType;
        }

        public void Draw(Graphics g)
        {
            Brush brush = new SolidBrush(Color.Red);
            g.FillEllipse(brush, Position.X - SIZE / 2, Position.Y - SIZE / 2, SIZE, SIZE);
        }

        public bool Contains(PointF point)
        {
            return (Math.Abs(point.X - Position.X) < SIZE / 2) && (Math.Abs(point.Y - Position.Y) < SIZE / 2);
        }
    }

    public class PropertyEditor : Form
    {
        private CapturePoint capturePoint;
        private ComboBox dropdown;
        private Button saveButton;

        public PropertyEditor(CapturePoint point)
        {
            this.capturePoint = point;
            this.Text = "Edit Capture Point";
            this.Size = new Size(200, 100);

            dropdown = new ComboBox { Left = 10, Top = 10, Width = 150 };
            dropdown.Items.AddRange(new string[] { "Default", "Image Scan", "Video Capture", "Depth Scan" });
            dropdown.SelectedItem = capturePoint.CaptureType;

            saveButton = new Button { Text = "Save", Left = 10, Top = 40 };
            saveButton.Click += SaveCaptureType;

            Controls.Add(dropdown);
            Controls.Add(saveButton);
        }

        private void SaveCaptureType(object sender, EventArgs e)
        {
            capturePoint.CaptureType = dropdown.SelectedItem.ToString();
            this.Close();
        }
    }

    public class PathData
    {
        public List<PathPoint> Path { get; set; } = new List<PathPoint>();
    }

    public class PathPoint
    {-
        public float X { get; set; }
        public float Y { get; set; }
        public string CaptureType { get; set; }

        public PathPoint(float x, float y, string captureType)
        {
            X = x;
            Y = y;
            CaptureType = captureType;
        }
    }
}
